components/brave_wallet/common/f4_jumble.cc

```diff
@@ -0,0 +1,165 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/common/f4_jumble.h"
+
+#include <algorithm>
+#include <memory>
+#include <string>
+
+#include "base/containers/span.h"
+#include "base/notreached.h"
+#include "brave/third_party/argon2/src/src/blake2/blake2.h"
+
+namespace brave_wallet {
+
+namespace {
+
+using Blake2bPersonalBytes = std::array<uint8_t, BLAKE2B_PERSONALBYTES>;
+
+// Sizes for Blake2b are defined here https://zips.z.cash/zip-0316#solution
+const size_t kMinMessageSize = 48u;
+const size_t kMaxMessageSize = 4184368u;
+const size_t kLeftSize = 64u;
+static_assert(kLeftSize <= BLAKE2B_OUTBYTES);
+
+Blake2bPersonalBytes GetHPersonalizer(uint8_t i) {
+  return {85, 65, 95, 70, 52, 74, 117, 109, 98, 108, 101, 95, 72, i, 0, 0};
+}
+
+Blake2bPersonalBytes GetGPersonalizer(uint8_t i, uint16_t j) {
+  return {85,
+          65,
+          95,
+          70,
+          52,
+          74,
+          117,
+          109,
+          98,
+          108,
+          101,
+          95,
+          71,
+          i,
+          static_cast<uint8_t>(j & 0xFF),
+          static_cast<uint8_t>(j >> 8)};
+}
+
+void FillBlake2bParamPersonal(const Blake2bPersonalBytes& personalizer,
+                              blake2b_param& params) {
+  CHECK_EQ(personalizer.size(), sizeof(params.personal));
+  memcpy(params.personal, personalizer.data(), sizeof(params.personal));
+}
+
+std::vector<uint8_t> blake2b(const std::vector<uint8_t>& payload,
+                             const Blake2bPersonalBytes& personalizer,
+                             size_t digest_len) {
+  CHECK(digest_len >= 1 && digest_len <= BLAKE2B_OUTBYTES);
+  blake2b_state blake_state = {};
+  blake2b_param params = {};
+
+  params.digest_length = digest_len;
+  params.fanout = 1;
+  params.depth = 1;
+
+  FillBlake2bParamPersonal(personalizer, params);
+
+  if (blake2b_init_param(&blake_state, &params) != 0) {
+    NOTREACHED_NORETURN();
+  }
+  if (blake2b_update(&blake_state, payload.data(), payload.size()) != 0) {
+    NOTREACHED_NORETURN();
+  }
+  std::vector<uint8_t> result(digest_len);
+  if (blake2b_final(&blake_state, result.data(), digest_len) != 0) {
+    NOTREACHED_NORETURN();
+  }
+
+  return result;
+}
+
+size_t GetLeftSize(const std::vector<uint8_t>& message) {
+  return std::min(kLeftSize, message.size() / 2);
+}
+
+std::vector<uint8_t> GetLeft(const std::vector<uint8_t>& message) {
+  return std::vector(message.begin(), message.begin() + GetLeftSize(message));
+}
+
+std::vector<uint8_t> GetRight(const std::vector<uint8_t>& message) {
+  return std::vector(message.begin() + GetLeftSize(message), message.end());
+}
+
+std::vector<uint8_t> h_round(uint8_t iter,
+                             const std::vector<uint8_t>& left,
+                             const std::vector<uint8_t>& right) {
```

Comment:
These would be better as `base::span<const uint8_t>`. That applies for all cases where we don't need to mutate the elements.

Reply:
Done